localstack/services/sns/provider.py

```diff
@@ -404,14 +404,33 @@ def confirm_subscription(
         authenticate_on_unsubscribe: authenticateOnUnsubscribe = None,
     ) -> ConfirmSubscriptionResponse:
         # TODO: validate format on the token (seems to be 288 hex chars)
-        store = self.get_store(account_id=context.account_id, region=context.region)
-        topic_tokens = store.subscription_tokens.get(topic_arn, {})
+        # this request can come from any http client, it might not be signed (we would need to implement
+        # `authenticate_on_unsubscribe` to force a signing client to do this request.
+        # so, the region and account_id might not be in the request. Use the ones from the topic_arn
+        try:
+            parsed_arn = parse_arn(topic_arn)
+        except InvalidArnException:
+            raise InvalidParameterException("Invalid parameter: Topic")
+
+        store = self.get_store(account_id=parsed_arn["account"], region=parsed_arn["region"])
+
+        # it seems SNS is able to know what the region of the topic should be, even though a wrong topic is accepted
+        if parsed_arn["region"] != get_region_from_subscription_token(token):
+            raise InvalidParameterException("Invalid parameter: Topic")
 
-        subscription_arn = topic_tokens.pop(token, None)
+        subscription_arn = store.subscription_tokens.get(token)
         if not subscription_arn:
             raise InvalidParameterException("Invalid parameter: Token")
 
         subscription = store.subscriptions.get(subscription_arn)
+        if not subscription:
```

Comment:
I should add a todo here. It might be that confirming a deleted subscription might actually recreate it, from the description of `UnsubscribeURL` messages. However, this is out of scope of the PR.